Add `webpack-config-flag-plugin`.

sgomes opened the pull request:
A new `webpack` plugin, that safely replaces `config.isEnabled` flag checks with literals, taking into account the various possible ways `isEnabled` can be imported.

It's being used to replace just the `desktop` config flag, initially, as that's the same across all (web) environments.

This work evolved from @jsnajdr's excellent explorations in #31587.

#### Changes proposed in this Pull Request

* Add `webpack-config-flag-plugin`.
* Add extensive tests for plugin.
* Drive-by fixes to `state/lib/middleware` in preparation for using the plugin with the `desktop` flag. Taken from #31587.
* Replace `desktop` config flag using plugin.

#### Testing instructions

* Ensure that the build works normally
* Ensure that the `async-load-jquery` chunk is not being built

sgomes committed:
Add `babel-plugin-config-flag`.

Replaces `config.isEnabled` flag checks with literals, safely.

sgomes committed:
Drive-by fixes to `state/lib/middleware`.

sgomes committed:
Fix example syntax highlighter type in readme.

sgomes committed:
Indent tests using tabs, not spaces.

ockham commented:
Will this allow us to drop [client-side config processing](https://github.com/Automattic/wp-calypso/tree/61621f8abdda7f7d4f67edaa6cfdeebb66b8c95e/client/config)?

sgomes commented:
> Will this allow us to drop [client-side config processing](https://github.com/Automattic/wp-calypso/tree/61621f8abdda7f7d4f67edaa6cfdeebb66b8c95e/client/config)?

The initial use-case is to replace client-side checks for the `desktop` flag only, plus maybe any flags that are the same across every (web) environment.

@blowery has discussed the possibility of moving to a multi-build setup, however, where we'd build for each environment individually, instead of having a single build for all environments. In that case, we could apply the plugin to every flag, instead of using a whitelist, and drop client-side config processing altogether.

sgomes committed:
Replace `desktop` flag using plugin.

sgomes commented:
This PR is now ready for review!

jsnajdr commented on babel.config.js:
```diff
@@ -1,14 +1,22 @@
+	plugins: [
+		[ '@automattic/transform-wpcalypso-async', { async: isBrowser && codeSplit } ],
+		[
+			'./packages/babel-plugin-config-flag',
```
Can you add the package to NPM dependencies and reference it as `@automattic/babel-plugin-config-flag`?

sgomes replied:
Done, with the `webpack` plugin.

jsnajdr reviewed (commented):
Thank you for picking up this old idea and running with it!

I think that this optimization still philosophically belongs to webpack rather than Babel and should be a webpack plugin.

Is the main reason for Babel that it provides API to work with the variable scope? webpack should be able to do it, too, because it's crucial for its main job: finding and replacing imports and exports. Webpack converts:
```js
import config from 'config';

config.isEnabled( 'desktop' );
```
to
```js
const configModule = webpackCodeForImportSideEffect( 'config' );

webpackCodeToGetDefaultImportFrom( configModule ).isEnabled( 'desktop' );
```

I.e., webpack needs to find all usages of the imported binding and replace it with its own code templates. That's almost exactly the same thing as replacing the `config.isEnabled` call.

sgomes committed:
Rewrite plugin as a webpack plugin.

It makes more sense to handle this at the webpack level, as pointed out
by @jsnajdr.

sgomes committed:
Use alias for plugin.

sgomes commented:
> I think that this optimization still philosophically belongs to webpack rather than Babel and should be a webpack plugin.

Thanks, @jsnajdr! After carefully studying some default plugins, I was able to find an undocumented way of achieving this, so I went ahead and rewrote the whole thing as a `webpack` plugin. It was a lot of work, but I think it does make more sense.

sgomes committed:
Update test snapshot.

sgomes committed:
Move some test comments around.

jsnajdr commented on packages/webpack-config-flag-plugin/index.js:
```diff
@@ -0,0 +1,100 @@
+					for ( const sp of specifiers ) {
+						// Default import (`import config from 'config'`)
+						if ( sp.type === 'ImportDefaultSpecifier' ) {
+							this.moduleName = sp.local && sp.local.name;
```
The `sp.local` object is always present, isn't it? The specifier always refers some local binding, otherwise it wouldn't be there.

sgomes replied:
Possibly? I tend to code rather defensively when dealing with undocumented APIs... Removed.

jsnajdr replied:
The AST API is standardized and documented in the ESTree spec. `ModuleSpecifier` and its `local` field is documented here: https://github.com/estree/estree/blob/master/es2015.md#modulespecifier

jsnajdr commented on packages/webpack-config-flag-plugin/index.js:
```diff
@@ -0,0 +1,100 @@
+			parser.hooks.import.tap( 'ConfigFlagPlugin', ( statement, source ) => {
+				if ( source === 'config' ) {
+					// We have an import of 'config'.
+					const specifiers = statement.specifiers || [];
```
Can return early here:
```js
if ( ! statement.specifiers ) return;
```

sgomes replied:
👍

jsnajdr commented on packages/webpack-config-flag-plugin/index.js:
```diff
@@ -0,0 +1,100 @@
+			parser.hooks.program.tap( 'ConfigFlagPlugin', () => {
+				// Reset imports on new file.
+				this.moduleName = undefined;
+				this.methodName = undefined;
```
Let's hope that future versions of webpack don't run the parser on multiple files in parallel 🙂 

I think webpack has `parser.state` to handle state like this safely.

sgomes replied:
Switched to using objects indexed by `parser.state.current.resource`.

jsnajdr commented on packages/webpack-config-flag-plugin/index.js:
```diff
@@ -0,0 +1,100 @@
+// and return the `flag` literal value.
+const isNamedCall = ( parser, methodName, expr ) => {
+	return methodName &&
+		parser.evaluate( methodName ).identifier === 'imported var' &&
```
This `parser.evaluate` call takes the currently walked AST node into account, and checks that the `config` variable refers to the imported binding and not some local scope override?

jsnajdr replied:
I spent some time now learning about this `imported var` stuff. It works, but there is an advance warning: webpack 5 tracks the imports completely differently -- there's no longer any `imported var` rename.

sgomes replied:
As far as I can tell from these undocumented APIs, `parser.evaluate` evaluates an expression in the current context. In this case, since we're giving it an identifier, the `identifier` property will be set, and will always be `'imported var'`. If there were something local overriding the import, this property would be set to something else (the name of the const, if I recall correctly).

We're building our castle on quicksand here, but such is life with `webpack` plugins 🤷‍♂

sgomes replied:
> I spent some time now learning about this `imported var` stuff. It works, but there is an advance warning: webpack 5 tracks the imports completely differently -- there's no longer any `imported var` rename.

Oh, interesting! Any chance you could point me to that documentation? It's hard to decide what to build on with undocumented APIs...

jsnajdr replied:
There's no documentation I'm aware of -- just the source.

jsnajdr replied:
> the identifier property will be set, and will always be 'imported var'

That's because one of the parser plugins that walks through `import` and `export` statements will set `parser.renames.set( 'config', 'imported var' )` and `parser.evaluateExpression` will respect these renames. That's why you get `imported var` as the identifier name. Webpack 5 doesn't do that rename any more. I don't understand yet what's going on there.

sgomes committed:
Remove unnecessary check.

sgomes committed:
Return early when missing specifiers.

sgomes committed:
Allow for concurrent processing.

jsnajdr reviewed (approved):
We'd need some more work to make this plugin perfect: use `parser.state`, prepare for webpack 5... But it's totally good enough to be merged and used :+1:

ockham commented:
This might actually come in handy for non-Calypso products such as Jetpack, where we've discussed (lack of) a config/feature flag system before (and where a purely compile-time system might be sufficient).

/cc @sirreal @simison @tyxla
